include/net/buf.h

```diff
@@ -1332,29 +1332,170 @@ int net_buf_linearize(void *dst, size_t dst_len,
  *         length if other timeout than K_FOREVER was used, and there
  *         were no free fragments in a pool to accommodate all data.
  */
-u16_t net_buf_append_bytes(struct net_buf *buf, u16_t len,
-			   const u8_t *value, s32_t timeout,
-			   net_buf_allocator_cb allocate_cb, void *user_data);
+u16_t net_buf_append_bytes(struct appendable_net_buf *abuf, u16_t len,
+			   const u8_t *value, s32_t timeout);
+
+/**
+ * @brief Delete existing fragment from a list of buffer fragments
+ *
+ * @param frags Pointer to a list of bugger fragments.
+ * @param parent parent fragment of frag, or NULL if none.
+ * @param frag Fragment to delete.
+ *
+ * @return Pointer to the following fragment, or NULL if it had no
+ *         further fragments.
+ */
+struct net_buf *net_buf_frag_del_ref(struct net_buf **frags,
+				     struct net_buf *parent,
+				     struct net_buf *frag);
+
+/**
+ * @brief Compact a list of buffer fragments.
+ *
+ * @details After this there is no more free space in individual fragments.
+ * @param frag Pointer to Network buffer fragment list.
+ *
+ * @return True if compact success, False otherwise.
+ */
+bool net_buf_compact(struct net_buf **frag);
+
+/**
+ * @brief  Insert data at an arbitrary offset in a series of fragments.
+ *
+ * @details Insert data to a net_buf. If there is not enough space in the
+ * net_buf then more net_buf will be added, unless there are no free net_buf
+ * and timeout occurs.
+ *
+ * @param abuf Appendable network buffer.
+ * @param frag   Network buffer fragment.
+ * @param offset Offset of frag where insertion will start.
+ * @param len Total length of input data
+ * @param data Data to be added
+ * @param timeout Timeout is passed to the appendable net_buf allocator
+ * callback.
+ *
+ * @return True on success, False otherwise.
+ */
+bool net_buf_insert_bytes(struct appendable_net_buf *abuf,
+			  struct net_buf *frag, u16_t offset,
+			  u16_t len, u8_t *data, s32_t timeout);
+
+/**
+ * @brief Get data from buffer
+ *
+ * @details Get N number of bytes starting from fragment's offset. If the total
+ * length of data is placed in multiple fragments, this function will read from
+ * all fragments until it reaches N number of bytes. Caller has to take care of
+ * endianness if needed.
+ *
+ * @param frag Network buffer fragment.
+ * @param offset Offset of input buffer.
+ * @param pos Pointer to position of offset after reading n number of bytes,
+ *            this is with respect to return buffer(fragment).
+ * @param len Total length of data to be read.
+ * @param data Data will be copied here.
+ *
+ * @return Pointer to the fragment or
+ *         NULL and pos is 0 after successful read,
+ *         NULL and pos is 0xffff otherwise.
+ */
+struct net_buf *net_buf_read(struct net_buf *frag, u16_t offset,
+			     u16_t *pos, u16_t len, u8_t *data);
+
+/**
+ * @brief Get a byte value from fragmented buffer
+ *
+ * @param frag Network buffer fragment.
```

Comment:
Copy-paste errors/confusion (?), missing opportunity to improve the API. E.g.: 

>offset Offset of input buffer

which "input buffer" and "offset **of**"??


> pos Pointer to position of offset

Position of offset? What's that, in plain words?

Why these are still 2 params, if at least 95% of the code has those 2 params as `pos, &pos`?

> after reading 2 bytes

Not 2, 1 here.

> value Value is returned

Value is returned?